### PR TITLE
chain: Patch to v0.38 with IBC

### PR DESCRIPTION
**Pull request by sorawit**
Now that phase 0 is release. It's time to go into the more fun territory.

Things already tested
- Send transactions with bandoracled reporting
- Work with docker-compose
- Work with db (blocks / transactions)

Things not done yet.
- Patch db to support new messages
- Integrate IBC
- Bring back proof gen

### CODE REVIEW

File: chain/Makefile [hunk 1]
```diff
@@ -17,7 +17,6 @@ all: lint install
 install: go.sum
 	go install -mod=readonly $(BUILD_FLAGS) ./cmd/bandd
 	go install -mod=readonly $(BUILD_FLAGS) ./cmd/bandcli
-	go install -mod=readonly $(BUILD_FLAGS) ./cmd/bandsv
```
Comment (smiled0g) [inline chain/Makefile:20]: this is no longer used?

File: chain/app/app.go [hunk 2]
```diff
@@ -5,111 +5,103 @@ import (
...
+	"github.com/cosmos/cosmos-sdk/x/upgrade"
+	upgradeclient "github.com/cosmos/cosmos-sdk/x/upgrade/client"
 
 	bandsupply "github.com/bandprotocol/bandchain/chain/x/supply"
 	"github.com/bandprotocol/bandchain/chain/x/zoracle"
 )
 
 const (
-	appName          = "BandApp"
+	AppName          = "BandApp"
```
Comment (smiled0g) [inline chain/app/app.go:44]: good change 👍

File: chain/app/app.go [hunk 3]
```diff
@@ -120,246 +112,233 @@ type bandApp struct {
...
 func NewBandApp(
 	logger log.Logger, db dbm.DB, traceStore io.Writer, loadLatest bool,
-	invCheckPeriod uint, baseAppOptions ...func(*bam.BaseApp),
+	invCheckPeriod uint, skipUpgradeHeights map[int64]bool, home string,
+	baseAppOptions ...func(*bam.BaseApp),
 ) *bandApp {
 
-	// First define the top level codec that will be shared by the different modules
-	cdc := MakeCodec()
+	// TODO: Remove cdc in favor of appCodec once all modules are migrated.
```
Comment (smiled0g) [inline chain/app/app.go:153]: so the `appCodec` is the new one that uses `Proto` instead of `amino`?

File: chain/app/db_app.go [hunk 4]
```diff
@@ -52,15 +54,16 @@ func (app *dbBandApp) InitChain(req abci.RequestInitChain) abci.ResponseInitChai
 	app.cdc.MustUnmarshalJSON(req.AppStateBytes, &genesisState)
 
 	// Genaccount genesis
-	var genaccountsState genaccounts.GenesisState
-	genaccounts.ModuleCdc.MustUnmarshalJSON(genesisState[genaccounts.ModuleName], &genaccountsState)
+	// TODO: Figure out how the genesis account is populated
```
Comment (smiled0g) [inline chain/app/db_app.go]: What does this mean? Would there be any difference in the genesis creation process and node mining #1 block?

File: chain/bandlib/client.go [hunk 5]
```diff
@@ -107,7 +107,7 @@ func (client *BandStatefulClient) loop() {
 					{
 						client.msgs = append(client.msgs, msg.msg)
 						client.totalGas += msg.gas
-						client.totalFees = client.totalFees.Add(msg.fees)
+						client.totalFees = client.totalFees.Add(msg.fees...)
```
Comment (smiled0g) [inline chain/bandlib/client.go:110]: Why does this need to be changed?

File: chain/client/client.go [hunk 6]
```diff
@@ -1,244 +1,244 @@
 package rpc
```
Comment (smiled0g) [inline chain/client/client.go:1]: Will this file be revisited & ported later? I'd be nice if you've commented on why the entire file is commented and what's the TODO for this.

File: chain/client/proof.go [hunk 7]
```diff
@@ -1,536 +1,536 @@
 package rpc
```
Comment (smiled0g) [inline chain/client/proof.go:1]: Same as above -- would be nice to have TODO here.

File: chain/client/proof_test.go [hunk 8]
```diff
@@ -1,88 +1,88 @@
 package rpc
```
Comment (smiled0g) [inline chain/client/proof_test.go:1]: TODO needed.

File: chain/client/validators.go [hunk 9]
```diff
@@ -1,64 +1,64 @@
 package rpc
```
Comment (smiled0g) [inline chain/client/validators.go:1]: TODO needed.

File: chain/cmd/bandcli/main.go [hunk 10]
```diff
@@ -105,23 +116,46 @@ func txCmd(cdc *amino.Codec) *cobra.Command {
...
+		flags.LineBreak,
 		multiDelegateCommand(cdc),
-		client.LineBreak,
+		flags.LineBreak,
 	)
 
 	// add modules' tx commands
 	app.ModuleBasics.AddTxCommands(txCmd, cdc)
 
+	// remove auth and bank commands as they're mounted under the root tx command
```
Comment (smiled0g) [inline chain/cmd/bandcli/main.go:134]: Hmm .. why add `auth` and `bank` commands in the first place if we're going to remove it here? Doesn't make sense and looks like a hack to me.

File: chain/docker-config/add-account.exp [hunk 11]
```diff
@@ -2,11 +2,7 @@ set user [lindex $argv 0]
 set pw [lindex $argv 1]
 set mnemonic [lindex $argv 2]
 
-spawn bandcli keys add $user --recover
-expect "Enter a passphrase to encrypt your key to disk:"
-send "$pw\r"
-expect "Repeat the passphrase:"
-send "$pw\r"
+spawn bandcli keys add $user --recover --keyring-backend test
```
Comment (smiled0g) [inline chain/docker-config/add-account.exp:5]: Does this mean the passphrase is no longer needed? If so, then what if the user used a keystone from v0.37, would that be compatible with v0.38 here?

File: chain/docker-config/generate-genesis.sh [hunk 12]
```diff
@@ -35,12 +35,11 @@ expect $DIR/add-account.exp \
     "smile stem oven genius cave resource better lunar nasty moon company ridge brass rather supply used horn three panic put venue analyst leader comic" \
 
 # add accounts to genesis
-bandd add-genesis-account $(bandcli keys show validator1 -a) 10000000000000uband
-bandd add-genesis-account $(bandcli keys show validator2 -a) 10000000000000uband
-bandd add-genesis-account $(bandcli keys show validator3 -a) 10000000000000uband
-bandd add-genesis-account $(bandcli keys show validator4 -a) 10000000000000uband
-
-bandd add-genesis-account $(bandcli keys show requester -a) 10000000000000uband
+bandd add-genesis-account validator1 10000000000000uband --keyring-backend test
```
Comment (smiled0g) [inline chain/docker-config/generate-genesis.sh:38]: what is the `test` here -- is it a flag/option or user-defined string that we hard-coded here?

File: chain/docker-config/single-validator/run.sh [hunk 13]
```diff
@@ -13,4 +13,4 @@ mv config_tmp.toml config.toml
 cd /zoracle/
 
 mkdir ~/.banddb
-bandd start --rpc.laddr tcp://0.0.0.0:26657 --with-db sqlite3:/root/.banddb/main.db
+bandd start --rpc.laddr tcp://0.0.0.0:26657 --with-db "postgres: host=172.18.0.88 port=5432 user=postgres dbname=postgres password=postgrespassword sslmode=disable"
```
Comment (smiled0g) [inline chain/docker-config/single-validator/run.sh:16]: Let's add `# TODO` to remind us to remove the user & password. Not sure if it's configured to be accessible via localhost only, but it doesn't look secure here.

File: chain/go.mod [hunk 14]
```diff
@@ -3,21 +3,30 @@ module github.com/bandprotocol/bandchain/chain
...
+	github.com/snikch/goodman v0.0.0-20171125024755-10e37e294daa
+	github.com/spf13/afero v1.2.2 // indirect
+	github.com/spf13/cobra v0.0.6
 	github.com/spf13/viper v1.6.2
 	github.com/stretchr/testify v1.5.1
 	github.com/tendermint/go-amino v0.15.1
-	github.com/tendermint/iavl v0.12.4
-	github.com/tendermint/tendermint v0.32.9
-	github.com/tendermint/tm-db v0.2.0
+	github.com/tendermint/tendermint v0.33.1-dev3
```
Comment (smiled0g) [inline chain/go.mod]: God bless `tendermint v0.33.1-dev3` 🙏👼

File: chain/x/zoracle/endblock.go [hunk 15]
```diff
@@ -84,7 +84,7 @@ func handleEndBlock(ctx sdk.Context, keeper Keeper) sdk.Result {
 		// Must never overflow because we already checked for overflow above with
 		// gasConsumed + request.ExecuteGas (which is >= gasUsed).
 		if overflow {
-			panic(sdk.ErrorGasOverflow{Descriptor: "ExecuteRequest"})
+			panic("zoracle::handleEndBlock: Gas overflow")
```
Comment (smiled0g) [inline chain/x/zoracle/endblock.go]: Is `sdk.ErrorGasOverflow` deprecated? What's the reason for this change?

File: docker-compose.yaml [hunk 16]
```diff
@@ -160,25 +158,6 @@ services:
     ports:
       - 80:80
 
-  code-server:
```
Comment (smiled0g) [inline docker-compose.yaml:163]: We no longer need this?

**Review by smiled0g — COMMENTED** [review]
This is without the IBC right? I don't see new messages or changes under report/request handler.

EDIT: Nvm, just saw the change log.

File: chain/Makefile (hunk 1 above)
Comment (sorawit) [reply chain/Makefile:20]: Yeah bandsv is not used now that the frontend can send transaction directly and there's a separate faucet process.

File: chain/app/db_app.go (hunk 4 above)
Comment (sorawit) [reply chain/app/db_app.go]: Yes. there are some changes. `genaccounts` module is deprecated and genesis accounts are written on `bank`, we will need to patch this code to work with that, but it's not critical

File: chain/bandlib/client.go (hunk 5 above)
Comment (sorawit) [reply chain/bandlib/client.go:110]: They change the signature. The function now accepts variadic arguments (arbitrary number of Coin arguments) vs the old one that accepts one slice (one array of Coins).

File: chain/client/proof_test.go (hunk 8 above)
Comment (sorawit) [reply chain/client/proof_test.go:1]: Sure.

File: chain/cmd/bandcli/main.go (hunk 10 above)
Comment (sorawit) [reply chain/cmd/bandcli/main.go:134]: Not my code. It's from them! https://github.com/cosmos/gaia/blob/ibc-alpha/cmd/gaiacli/main.go#L134 

Didn't dig more, but I agree it looks hacky, but it must be here for some reasonss.

File: chain/docker-config/add-account.exp (hunk 11 above)
Comment (sorawit) [reply chain/docker-config/add-account.exp:5]: They change it so that there are multiple ways you can store your key. If user wants to use the same one as 0.37, they pass `--keyring-backend file`. The default option now uses the machine's keychain store.

File: chain/docker-config/generate-genesis.sh (hunk 12 above)
Comment (sorawit) [reply chain/docker-config/generate-genesis.sh:38]: It's the configuration of where to store & encrypted key (mnemonic).

File: chain/docker-config/single-validator/run.sh (hunk 13 above)
Comment (sorawit) [reply chain/docker-config/single-validator/run.sh:16]: This docker thing is primarily for testing.

File: chain/go.mod (hunk 14 above)
Comment (sorawit) [reply chain/go.mod]: lol

File: chain/x/zoracle/endblock.go (hunk 15 above)
Comment (sorawit) [reply chain/x/zoracle/endblock.go]: It's not. I did the change because I ran into some errors. Not sure why. But I change it back now and it works, so good catch!

File: chain/app/app.go (hunk 3 above)
Comment (sorawit) [reply chain/app/app.go:153]: It's a mess but yes. `appCodec` is their new codec effort.

File: docker-compose.yaml (hunk 16 above)
Comment (sorawit) [reply docker-compose.yaml:163]: It's the `bandsv` thing. No longer needed.

File: chain/app/db_app.go (hunk 4 above)
Comment (sorawit) [reply chain/app/db_app.go]: From Cosmos change log:
Screen Shot 2020-04-05 at 00 50 34

File: chain/bandlib/client.go (hunk 5 above)
Comment (sorawit) [reply chain/bandlib/client.go:110]: Screen Shot 2020-04-05 at 00 59 56

File: chain/app/db_app.go (hunk 4 above)
Comment (sorawit) [reply chain/app/db_app.go]: Nevermind. I patched it now. Tested locally and seemed to be working.

File: chain/app/genesis.go [hunk 17]
```diff
@@ -5,119 +5,67 @@ import (
...
-// GenesisState defines a type alias for the Gaia genesis application state.
+// GenesisState defines a type alias for the Band genesis application state.
 type GenesisState map[string]json.RawMessage
 
 // NewDefaultGenesisState generates the default state for the application.
 func NewDefaultGenesisState() GenesisState {
+	cdc := codecstd.MakeCodec(ModuleBasics)
+	denom := "uband"
+
+	stakingGenesis := staking.DefaultGenesisState()
```
Comment (taobun) [inline chain/app/genesis.go:36]: Prefer default instead of custom all fileds?

File: chain/db/validator.go [hunk 18]
```diff
@@ -116,9 +116,14 @@ func (b *BandDB) handleMsgEditValidator(msg staking.MsgEditValidator) error {
 }
 
 func (b *BandDB) handleMsgCreateValidator(msg staking.MsgCreateValidator) error {
+	pubKey, err := sdk.GetPubKeyFromBech32(sdk.Bech32PubKeyTypeConsPub, msg.Pubkey)
```
Comment (taobun) [inline chain/db/validator.go:119]: Wow pubkey change to string ...

File: chain/x/supply/keeper.go [hunk 19]
```diff
@@ -47,19 +49,24 @@ func (k WrappedSupplyKeeper) BurnCoins(
 	// Create the account if it doesn't yet exist.
 	acc := k.GetModuleAccount(ctx, moduleName)
 	if acc == nil {
-		return sdk.ErrUnknownAddress(fmt.Sprintf("module account %s does not exist", moduleName))
+		panic(sdkerrors.Wrapf(
```
Comment (taobun) [inline chain/x/supply/keeper.go:52]: Panic?

File: chain/x/zoracle/handler.go [hunk 20]
```diff
@@ -6,11 +6,12 @@ import (
 	"github.com/bandprotocol/bandchain/chain/owasm"
 	"github.com/bandprotocol/bandchain/chain/x/zoracle/internal/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )
 
 // NewHandler creates the msg handler of this module, as required by Cosmos-SDK standard.
 func NewHandler(keeper Keeper) sdk.Handler {
-	return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
+	return func(ctx sdk.Context, msg sdk.Msg) (*sdk.Result, error) {
```
Comment (taobun) [inline chain/x/zoracle/handler.go:14]: Return pointer???

File: chain/x/zoracle/internal/keeper/querier_test.go [hunk 21]
```diff
@@ -176,7 +176,7 @@ func TestQueryRequestById(t *testing.T) {
 		abci.RequestQuery{},
 	)
 	// It must return error request not found
-	require.Equal(t, types.CodeItemNotFound, err.Code())
+	require.NotNil(t, err)
```
Comment (taobun) [inline chain/x/zoracle/internal/keeper/querier_test.go:179]: Cannot test error type?

File: chain/app/app.go [hunk 22]
```diff
@@ -395,6 +374,32 @@ func (app *bandApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDeliv
...
+// InitChainer application update at chain initialization
+func (app *bandApp) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
+	var genesisState simapp.GenesisState
+	app.cdc.MustUnmarshalJSON(req.AppStateBytes, &genesisState)
+
+	res := app.mm.InitGenesis(ctx, app.cdc, genesisState)
+
+	// Set Historical infos in InitChain to ignore genesis params
+	stakingParams := staking.DefaultParams()
+	stakingParams.HistoricalEntries = 1000
```
Comment (taobun) [inline chain/app/app.go:396]: db app things? Have related logic in our chain?

File: chain/app/genesis.go (hunk 17 above)
Comment (sorawit) [reply chain/app/genesis.go:36]: Yes. I think this is safer. We overwrite exactly what we need and we leave other things with default values.

File: chain/db/validator.go (hunk 18 above)
Comment (sorawit) [reply chain/db/validator.go:119]: Yes

File: chain/x/supply/keeper.go (hunk 19 above)
Comment (sorawit) [reply chain/x/supply/keeper.go:52]: Yes. I'm making the behavior as similar to the original `BurnCoins` as possible.
https://github.com/cosmos/cosmos-sdk/blob/ibc-alpha/x/supply/keeper/bank.go#L130

File: chain/x/zoracle/handler.go (hunk 20 above)
Comment (sorawit) [reply chain/x/zoracle/handler.go:14]: Yes. The new spec of handler returns `(*sdk.Result, error)`

File: chain/x/zoracle/internal/keeper/querier_test.go (hunk 21 above)
Comment (sorawit) [reply chain/x/zoracle/internal/keeper/querier_test.go:179]: Yes. I can confirm that ALL tests in Cosmos-SDK use this approach. No tests check for error code.

File: chain/app/app.go (hunk 22 above)
Comment (sorawit) [reply chain/app/app.go:396]: This is not db app things. It's from here https://github.com/cosmos/gaia/blob/ibc-alpha/app/app.go#L328. 

Now that I look at it again. It's `gaia`'s IBC hack to store old validator set. I think we should leave it here for now. We can move it to default genesis param later. We can talk more if you are still not convinced.

**Review by taobun — APPROVED** [review]

File: chain/cmd/bandd/init.go [hunk 23]
```diff
@@ -162,17 +163,12 @@ func InitCmd(
...
-				},
-				Evidence: types.EvidenceParams{
-					MaxAge: 100000,
-				},
-				Validator: types.ValidatorParams{PubKeyTypes: []string{types.ABCIPubKeyTypeSecp256k1}},
-			}
+			genDoc.ConsensusParams = types.DefaultConsensusParams()
+			genDoc.ConsensusParams.Block.MaxBytes = 200000 // 0.2MB
+			genDoc.ConsensusParams.Block.MaxGas = 20000000 // 20M gas (Maximum oracle script size uses 15M)
+			genDoc.ConsensusParams.Block.TimeIotaMs = 1000 // 0.2MB
```
Comment (Benzbeeb) [inline chain/cmd/bandd/init.go:169]: I think it should be 1s

**Review by Benzbeeb — CHANGES_REQUESTED** [review]

**Comment by sorawit** [comment]
It's getting real!
